live_data_subscriber/websocket/test_scripts/ws_server.py

```diff
@@ -0,0 +1,46 @@
+import asyncio
+import logging
+import websockets
+from websockets import WebSocketServerProtocol
+
+logging.basicConfig(level=logging.INFO)
+
+
+class Server:
+    clients = set()
+
+    async def register(self, ws: WebSocketServerProtocol) -> None:
+        self.clients.add(ws)
+        logging.info(f'{ws.remote_address} connects.')
+
+    async def unregister(self, ws: WebSocketServerProtocol) -> None:
+        self.clients.remove(ws)
+        logging.info(f'{ws.remote_address} disconnects.')
+
+    async def send_to_clients(self, message: str) -> None:
+        if self.clients:
+            await asyncio.wait([client.send(message) for client in self.clients])
+
+    async def ws_handler(self, ws: WebSocketServerProtocol, uri: str) -> None:
+        await self.register(ws)
+        try:
+            await self.distribute(ws)
+        finally:
+            await self.unregister(ws)
+
+    async def distribute(self, ws: WebSocketServerProtocol) -> None:
+        async for message in ws:
+            await self.send_to_clients(message)
+
+
+def start_ws_server():
+    logging.info(f"start_ws_server")
+    server = Server()
+    start_server = websockets.serve(server.ws_handler, 'localhost', 4000)
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(start_server)
+    loop.run_forever()
```

Comment:
Aren't these two different handlers? Or can both functionalities be combined?
Copy/Paste from the asyncio.events:
```
    def run_forever(self):
        """Run the event loop until stop() is called."""
        raise NotImplementedError

    def run_until_complete(self, future):
        """Run the event loop until a Future is done.

        Return the Future's result, or raise its exception.
        """
        raise NotImplementedError
```

Reply:
After testing it both were needed. Since Im implementing it for the first time and not yet an expert so may not have the best answer for your question, `run_until_complete `  is an asynchronous iterator that do the job for objects being passed whereas `run_forever ` is to hold the script from exiting, so you would see `run_forever ` in `server` and `consumer´ script but not in `producer` script.